### ISSUE TITLE
Incorrect resolution of address of (&) operator

### ISSUE THREAD

**Issue by hasantouma** [microsoft/checkedc-clang#635]
The following code should not produce an error:

`deref.c`:
```
#include<stdio_checked.h>
int main() {
   _Nt_array_ptr<char> hel:count(4) = "hello";
   _Nt_array_ptr<char> h = NULL;
   h = &hel[2];
}
```

But I get this error:
```
deref.c:6:6: error: assigning to '_Nt_array_ptr<char>' from incompatible type 'char *'
   h = &hel[2];
     ^ ~~~~~~~
1 error generated.
```

**Comment by dtarditi** [comment]
In this specific case, you can work around the error by using `h = hel + 2`.    

This case highlighst the issue that the address-of operator (`&`) does not work well with `_Nt_array_ptr`.  The type of `&` is never an `_Nt_array_ptr` and there is no implicit conversion that allows its result to be converted to an `_Nt_array_ptr`.  As you point out, this leads to unexpected failures when converting code.

I think we can extend the existing implicit conversion rules and the inference of bounds for expressions to make the cases seen in conversion work. During inference of bounds expressions for expressions, we can track whether the upper bound is actually  null-terminated.  I considered changing the typing rules for `&`, but I don't think that works out well.

## Existing Checked C rules

In unchecked scopes, the `&` operator produces an unchecked pointer type.  Given an expression `e` of type `T` that is an lvalue expression, `&e` has type `T *`.   

At assignments, we allow implicit conversions from `T *` to `_Ptr<T>` or `_Array_ptr<T>` types.  This is with additional conditions around bounds.  For example, for `_Ptr`, the pointer being assigned must have bounds that imply the validity of the `_Ptr<T>`. However, we do not allow an implicit conversion from `T *` to `_Nt_array_ptr<T>` types.  In general, we don't know if the pointed-to array is null-terminated.

In checked scopes, `&e` produces an `_Array_ptr<T>`.   This rule also has problems for `_Nt_array_ptr` types, because we don't allow an `_Array_ptr<T>` to be converted to an `_Nt_array_ptr<T>`.

## Problems with changing typing rules for `&`

Given `&e`, `e` must be an lvalue expression. We could track whether `e` has an lvalue that is "derived" from a checked pointer and alter the typing rules for `&` based on that.  For example `&(*p)` could produce a checked pointer type if `p` is checked pointer.

This breaks down for some simple cases, though.  Consider:
```
int v;
&v;
```
Should `&v` have checked or unchecked pointer type?   In unchecked scopes, if you give it a checked pointer type, this will cause existing C code to not type check.

Now, you might say that existing rules cover this, we can just rely on implicit conversions from unchecked pointers to checked pointers.

Consider the case:
```
&("abcdef"[i]).
```
Should this produce `char *` or `_Array_ptr<char>`?  In unchecked scopes, "abcdef" must have unchecked pointer type.  Otherwise, existing code will break.  This means that the lvalue expression is not derived from a checked pointer and that the type of the entire expression would be `char *`.

## Proposed solution

We infer bounds expressions for expressions.  For any expression, we compute a bounds expression.  The result could be unknown bounds, any bounds, or bounds(lb, ub).   We can extend the algorithm to track whether the upper bounds is null-terminated or not.   At assignments, we can allow conversions from T * or _Array_ptr<T> to _Nt_array_ptr<T>, with the additional side condition that the upper bound must be null-terminated.

I'll sketch out a few examples:
- &v, where v is variable of scalar type, has bounds (&v, &v + 1). The 
upper bound is not null-terminated.
- e1 + e2, where e1 has pointer type, has a null-terminated upper bound if and only if
  e1 has a null-terminated upper bound
- v, where v is a pointer-typed variable, has a null-terminated upper bound if v has
  type _Nt_array_ptr.
- For a string literal s, we introduce a temporary t for the result of the string literal.  If the string literal has n characters in it (excluding the null terminator), the bounds are (t, t + n) and the upper bound is null-terminated. For example, for "abcdef", the bounds are (t, t + 6) and the upper bound is
null-terminated.
